test/e2e/e2e_test.go

```diff
@@ -0,0 +1,286 @@
+		t.Log("Expected:")
+		_ = e.Encode(event)
+		t.Log("Have:")
+		_ = e.Encode(agentMock.Events())
+		t.Fatalf("Event was not captured")
+	}
+}
+
+func TestPodDeletion(t *testing.T) {
+	client, agentMock := initialize(t)
+
+	t.Log("Creating test namespace...")
+	ns, err := client.CoreV1().Namespaces().Create(context.Background(), &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "e2e-" + strings.ToLower(t.Name()) + "-",
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		t.Logf("could not create %s namespace: %v", ns, err)
+	}
+
+	defer func() {
+		t.Log("Cleaning up test namespace...")
+		_ = client.CoreV1().Namespaces().Delete(context.Background(), ns.Name, metav1.DeleteOptions{})
+	}()
+
+	t.Log("Creating test pod...")
+	testpod, err := client.CoreV1().Pods(ns.Name).Create(context.Background(), &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "nginx-e2e-killable",
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "nginx",
+					Image: "nginx",
+				},
+			},
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		t.Errorf("could not create test pod: %v", err)
+		return
+	}
+
+	time.Sleep(7 * time.Second)
+
+	err = client.CoreV1().Pods(ns.Name).Delete(context.Background(), testpod.Name, metav1.DeleteOptions{})
+	if err != nil {
+		t.Errorf("could not create test pod: %v", err)
+		return
+	}
+
+	t.Log("Waiting for events to show up...")
+	agentMock.Wait(15*time.Second, 1*time.Minute)
+	for _, event := range []sdkEvent.Event{
+		{
+			Summary:  "Stopping container " + testpod.Spec.Containers[0].Name,
+			Category: "kubernetes",
+			Attributes: map[string]interface{}{
+				"event.metadata.name":             testpod.Name + ".",
+				"event.metadata.namespace":        ns.Name,
+				"event.reason":                    "Killing",
+				"clusterName":                     "",
+				"event.involvedObject.apiVersion": "",
+				"event.involvedObject.kind":       "Pod",
+				"event.involvedObject.name":       testpod.Name,
+				"event.message":                   "Stopping container " + testpod.Spec.Containers[0].Name,
+				"event.type":                      "Normal",
+				"verb":                            "ADDED",
+			},
+		},
+	} {
+		if agentMock.Has(&event) {
+			continue
+		}
+
+		e := json.NewEncoder(os.Stderr)
+		t.Log("Expected:")
+		e.Encode(event) // nolint:errcheck
+		t.Fatalf("Event was not captured")
+	}
+}
+
+// initialize returns a kubernets client and a mocked agent sink ready to receive events
+func initialize(t *testing.T) (*kubernetes.Clientset, *e2e.MockedAgentSink) {
+	t.Helper()
+
+	conf, err := restConfig()
+	if err != nil {
+		t.Fatalf("could not build kubernetes config: %v", err)
+	}
+
+	client, err := kubernetes.NewForConfig(conf)
+	if err != nil {
+		t.Fatalf("could not build kubernetes client: %v", err)
+	}
+
+	sharedInformers := informers.NewSharedInformerFactory(client, time.Duration(0))
+	eventsInformer := sharedInformers.Core().V1().Events().Informer()
```

Comment:
Maybe it would make sense to change the informer to only fetch events from a test namespace to avoid polluting it? This should also allow dropping the delete code below.

Reply:
This code is mostly copy-pasted from the `main` (since I cannot import it), so I wanted it to be as close as possible as the original.
Scoping events to namespaces doesn't sound bad, but I like the idea of creating one namespace per test case run and adding the filter here would not play well with that, right?

Reply:
Yes. You really can't create multiple sinks? That seems like a very big limitation.

Reply:
It is possible to create multiple sinks, but not multiple instances of the `InfrastructureAgentSink`. This is because it uses the [infra-integrations-sdk](https://github.com/newrelic/infra-integrations-sdk/), which is meant to mostly for CLIs, and makes calls to the shared instance of `flag`. If I attempt to create more than one, `flag` will panic the second time the infra-sdk attempts to register an already registered flag :(

Needless to say I dislike singletons with all the rage of my burning heart, but seems quite difficult to avoid here (specially since it's "just test code").

The SDK could be improved to just avoid doing this, but it's a big thing that would fall out of the scope of this epic.